packages/contractkit/src/wrappers/BaseWrapper.ts

```diff
@@ -46,3 +29,88 @@ export function toBigNumber(input: string) {
 export function toNumber(input: string) {
   return parseInt(input, 10)
 }
+
+export function parseNumber(input: string | number | BigNumber) {
+  return new BigNumber(input).toString()
+}
+
+type Parser<A, B> = (input: A) => B
+
+export function tupleParser<A0, B0>(parser0: Parser<A0, B0>): (...args: [A0]) => [B0]
+export function tupleParser<A0, B0, A1, B1>(
+  parser0: Parser<A0, B0>,
+  parser1: Parser<A1, B1>
+): (...args: [A0, A1]) => [B0, B1]
+export function tupleParser<A0, B0, A1, B1, A2, B2>(
+  parser0: Parser<A0, B0>,
+  parser1: Parser<A1, B1>,
+  parser2: Parser<A2, B2>
+): (...args: [A0, A1, A2]) => [B0, B1, B2]
+export function tupleParser<A0, B0, A1, B1, A2, B2, A3, B3>(
+  parser0: Parser<A0, B0>,
+  parser1: Parser<A1, B1>,
+  parser2: Parser<A2, B2>,
+  parser3: Parser<A3, B3>
+): (...args: [A0, A1, A2, A3]) => [B0, B1, B2, B3]
+export function tupleParser(...parsers: Array<Parser<any, any>>) {
+  return (...args: any[]) => zip((parser, input) => parser(input), parsers, args)
+}
+
+type ProxyCallArgs<A extends any[], B extends any[], C, D> =
+  | [Method<B, C>, (...arg: A) => B, (arg: C) => D]
+  | [Method<A, C>, undefined, (arg: C) => D]
+  | [Method<B, D>, (...arg: A) => B]
+  | [Method<A, D>]
+
+export function proxyCall<A extends any[], B extends any[], C, D>(
```

Comment:
Could you add some comments explaining what this does? I think sometimes we do have to do some ugly looking things, but it's worth explaining why so then we can constrain it to one spot

Reply:
I'll add some comments.

I don't think this is ugly, but it's a complex type. Not an easy think to type.

The intention here is to provide a function that can receive a preParser or a postParser or both or none.. and still provide a good return type.. in this case.. it's always `(...args: A) => Promise<D>` no matter the combination.

I've spent quite some time trying to type this, and avoid the `ts-ignore` or `any`.

But open to suggestions if you know a better way